strutil/limbuffer.go

```diff
@@ -0,0 +1,61 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+import (
+	"bytes"
+)
+
+type LimitedBuffer struct {
+	buffer   *bytes.Buffer
+	maxLines int
+	maxBytes int
+	written  int
+}
+
+func NewLimitedBuffer(maxLines, maxBytes int) *LimitedBuffer {
+	return &LimitedBuffer{
+		buffer:   &bytes.Buffer{},
+		maxLines: maxLines,
+		maxBytes: maxBytes,
+	}
+}
+
+func (wr *LimitedBuffer) Write(data []byte) (int, error) {
+	sz, err := wr.buffer.Write(data)
+	if err != nil {
+		return 0, err
+	}
+	wr.written += sz
+	if wr.written > wr.maxBytes {
+		data := wr.buffer.Bytes()
```

Comment:
BTW. I'm correctly following this, the actual storage i.e. []byte in buffer is not getting reallocated and we're in fact only sliding down the data.

On the other hand, if `Write()` i scalled with `data` that is a many times larger than `maxBytes`, the internal buffer will grow to `wr.buffer.Len()` + that size. Probably a bit of an overkill but I think the truncation should be done by sliding down as much of the old data as possible (given `maxBytes - length of new data`, maybe even none if `len(data) > maxBytes`) and appending/copying the new data right after the old.